src/linear_algebra/eigenspace.lean

```diff
@@ -0,0 +1,231 @@
+`x ∈ eigenspace f μ`.
+
+## Notations
+
+The expression `algebra_map K (End K V)` appears very often, which is why we use `am` as a local
+notation for it.
+
+## References
+
+* [Sheldon Axler, *Down with determinants!*,
+  https://www.maa.org/sites/default/files/pdf/awards/Axler-Ford-1996.pdf][axler1996]
+* https://en.wikipedia.org/wiki/Eigenvalues_and_eigenvectors
+
+## Tags
+
+eigenspace, eigenvector, eigenvalue, eigen
+-/
+
+universes u v w
+
+open vector_space principal_ideal_ring polynomial finite_dimensional
+
+variables {K : Type v} {V : Type w} [add_comm_group V]
+
+local notation `am` := algebra_map K (module.End K V)
+
+/-- The subspace `eigenspace f μ` for a linear map `f` and a scalar `μ` consists of all vectors `x`
+    such that `f x = μ • x`. -/
+def eigenspace [comm_ring K] [module K V] (f : module.End K V) (μ : K) : submodule K V :=
+(f - am μ).ker
+
+/-- A nonzero element of an eigenspace is an eigenvector. -/
+def eigenvector [comm_ring K] [module K V] (f : module.End K V) (μ : K) (x : V) : Prop :=
+x ≠ 0 ∧ x ∈ eigenspace f μ
+
+/-- A scalar `μ` is an eigenvalue for a linear map `f` if there are nonzero vectors `x`
+    such that `f x = μ • x`. -/
+def eigenvalue [comm_ring K] [module K V] (f : module.End K V) (a : K) : Prop :=
+eigenspace f a ≠ ⊥
+
+lemma mem_eigenspace_iff [comm_ring K] [module K V]
+ {f : module.End K V} {μ : K} {x : V} : x ∈ eigenspace f μ ↔ f x = μ • x :=
+by rw [eigenspace, linear_map.mem_ker, linear_map.sub_apply, module.algebra_map_End_apply,
+   sub_eq_zero]
+
+lemma eigenspace_div [field K] [vector_space K V] (f : module.End K V) (a b : K) (hb : b ≠ 0) :
+  eigenspace f (a / b) = (b • f - am a).ker :=
+calc
+  eigenspace f (a / b) = eigenspace f (b⁻¹ * a) : by { dsimp [(/)], rw mul_comm }
+  ... = (f - (b⁻¹ * a) • linear_map.id).ker : rfl
+  ... = (f - b⁻¹ • a • linear_map.id).ker : by rw smul_smul
+  ... = (f - b⁻¹ • am a).ker : rfl
+  ... = (b • (f - b⁻¹ • am a)).ker : by rw linear_map.ker_smul _ b hb
+  ... = (b • f - am a).ker : by rw [smul_sub, smul_inv_smul' hb]
+
+lemma eigenspace_eval₂_polynomial_degree_1 [field K] [vector_space K V]
+  (f : module.End K V) (q : polynomial K) (hq : degree q = 1) :
+  eigenspace f (- q.coeff 0 / q.leading_coeff) = (eval₂ am f q).ker :=
+calc
+  eigenspace f (- q.coeff 0 / q.leading_coeff) = (q.leading_coeff • f - am (- q.coeff 0)).ker
+    : by { rw eigenspace_div, intro h, rw leading_coeff_eq_zero_iff_deg_eq_bot.1 h at hq, cases hq }
+  ... = (eval₂ am f (C q.leading_coeff * X + C (q.coeff 0))).ker
+    : by { rw C_mul', simpa [algebra_map, algebra.to_ring_hom] }
+  ... = (eval₂ am f q).ker
+     : by { congr, apply (eq_X_add_C_of_degree_eq_one hq).symm }
+
+/-- Every linear operator on a vector space over an algebraically closed field has
+    an eigenvalue. (Axler's Theorem 2.1.) -/
+lemma exists_eigenvalue
+  [field K] [is_alg_closed K] [vector_space K V] [finite_dimensional K V]
+  (f : module.End K V) (hex : ∃ v : V, v ≠ 0) :
```

Comment:
Is `hex` the canonical way to say this, or would this be better as `dim K V > 0`?

Reply:
`[nontrivial v]`